torch/csrc/autograd/engine.cpp

```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
   std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
 }
 
-void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
+std::shared_ptr<FutureVariableList> Engine::execute_graph_task_until_ready_queue_empty(
+    const std::shared_ptr<GraphTask>& graph_task,
+    std::shared_ptr<Node> root_to_execute,
+    bool incrementOutstandingTasks) {
   initialize_device_threads_pool();
-  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
-  // The graph_task must be alive at this point, because internal autograd machinary
-  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
-  // outside this API to keep the GraphTask alive
-  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
-  ready_queue(graph_task, at::kCPU)->push(
-      std::move(task), /* incrementOutstandingTasks */ false);
-}
-
-void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
   std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
```

Comment:
If we share one ready queue for the whole graph task, everytime a new backward is called, we create a new worker which can take nodes from the same readyqueue other backward calls populated. Our general threading strategy is fork and rpc in the forward path makes the corresponding backward  parallelizable, and ops in the corresponding "fork" and "rpc" should be run in the same thread. The behavior here makes backward and forward unsymmetric. It will be much harder to reason and explain.

Reply:
definitely.